### DISCUSSION TITLE
How to create browser history in SPA

### DISCUSSION THREAD

**Discussion by RamKapil** [quasarframework/quasar#11185] [General - Components / Directives / etc] [answered]
I am creating an SPA.  routing works for my application but it does not create browser history.  As per Vue documentation, I am supposed to call createWebHistory() function when creating router object.  However, in Quasar we don't explicitly create a router object.  Please help me understand how browser history can be created in SPAs so that when user clicks "back" button on browser, it takes the user to the previous page.

In quasar.conf.js, I have set vueRouterMode: "history".  it removes the '#' but does not create browser history.  Is there any other setting which can solve my problem.

**Comment by MilosPaunovic** [comment]
`vueRouterMode: 'history'` should remove the hash from the URL, but the browsers back button should work out of the box. What happens when you click back, do you have problems only in your local environment or with built and deployed code?

**Reply by RamKapil** [reply]
I have not deployed this application yet.  http://localhost:8080/ is the base url.  after that I navigate to http://localhost:8080/charts by clicking a button in the app.  Now, when I click "back" button of browser, it does not do anything.  it stays on http://localhost:8080/charts,  thanks for your prompt reply.

Just noticed that "http://localhost:8080/" is the only address that is not getting added to browser history.  Opened a fresh instance of browser, navigated to http://localhost:8080/.... then http://localhost:8080/charts.  Now, when clicked back browser button, it took me back to blank browser page.. skipping http://localhost:8080/.

**Comment by MilosPaunovic** [comment]
If you enter `http://localhost:8080/charts` into a fresh instance of browser, back button is supposed to take you back to blank browser page. If you enter ` http://localhost:8080`, then navigate to `/charts` via `vue-router` link, than history stack is being filled and you can use back button to go back to previous page.

**Reply by RamKapil** [reply]
yes.  http://localhost:8080, then navigate to /charts via vue-router, but it takes to blank page when browser back button is clicked.  following is the route declaration, if that can help identify the problem. for / and /charts, I am using separate layout:

const routes = [
  {
    path: "/",
    component: () => import("layouts/MainLayout.vue"),
    children: [{ path: "/", component: () => import("pages/Index.vue") }],
  },
  {
    path: "/charts",
    component: () => import("layouts/ChartLayout.vue"),
    children: [
      { path: "/charts", component: () => import("pages/Charts.vue") },
    ],
  },

**Reply by MilosPaunovic** [reply]
Try to remove everything from children path attributes because it is the same as parent route path and see if that does the job.

```js
children: [{ path: "", component: () => import("pages/Index.vue") }],
children: [{ path: "", component: () => import("pages/Charts.vue") }]
```

**Reply by RamKapil** [reply]
tried.  It doesn't solve the problem.

**Reply by RamKapil** [reply]
changed routes for both pages to use same layout, but still no luck -
const routes = [
  {
    path: "",
    component: () => import("layouts/MainLayout.vue"),
    children: [
      { path: "", component: () => import("pages/Index.vue") },
      { path: "/charts", component: () => import("pages/Charts.vue") },
    ],
  },

**Reply by MilosPaunovic** [reply]
Could you provide a reproduction repo so someone could investigate further?

**Reply by RamKapil** [reply]
do you mean source code?.  ok. i will upload it on github.

**Reply by RamKapil** [reply]
The problem is resolved.  I am quite new to Vue js and Quasar.  I was using - @click="$router.replace('/second')" with buttons for navigation.  I now found that router-link component has to be used instead.   With this change, browser history is being created and back button now takes me to the previous page.   Thanks you very much for all your time and support.

**Comment by RamKapil** [comment, marked as answer]
The problem is resolved. I am quite new to Vue js and Quasar. I was using - @click="$router.replace('/second')" with buttons for navigation. I now found that "router-link component " (https://router.vuejs.org/api/#router-link)  has to be used instead. With this change, browser history is being created and back button now takes me to the previous page.

**Comment by rstoenescu** [comment]
> However, in Quasar we don't explicitly create a router object.

Yes we do! Check /src/router/index.js/ts
